terps/unp64/unp64.cpp

```diff
@@ -0,0 +1,745 @@
+				_G(_unp)._strMem = 2;
+			break;
+		}
+
+		_G(_iter)++;
+		if (_G(_iter) == iterMax) {
+			return 0;
+		}
+
+		if ((r->_pc >= 0xa000) && (r->_pc <= 0xbfff) && ((mem[1] & 0x7) == 7)) {
+			if (isBasicRun2(r->_pc)) {
+				r->_pc = 0xa7ae;
+				break;
+			} else {
+				mem[0] = 0x60;
+				r->_pc = 0;
+			}
+		}
+
+		if (r->_pc >= 0xe000) {
+			if (((mem[1] & 0x7) >= 6) && ((mem[1] & 0x7) <= 7)) {
+				if (r->_pc == 0xffbd) {
+					mem[0xB7] = r->_a;
+					mem[0xBB] = r->_x;
+					mem[0xBC] = r->_y;
+				}
+				/* return into IRQ handler, better stop here */
+				if (((r->_pc >= 0xea31) && (r->_pc <= 0xeb76)) || (r->_pc == 0xffd5) || (r->_pc == 0xfce2)) {
+					break;
+				}
+
+				if (r->_pc == 0xfda3) {
+					mem[0x01] = 0xe7;
+					r->_a = 0xd7;
+					r->_x = 0xff;
+				}
+				mem[0] = 0x60;
+				r->_pc = 0;
+			}
+		}
+	}
+
+	if (_G(_unp)._fEndAf && _G(_unp)._monEnd) {
+		_G(_unp)._endAdC = mem[_G(_unp)._fEndAf] | mem[_G(_unp)._fEndAf + 1] << 8;
+		if ((int)_G(_unp)._endAdC > _G(_unp)._endAdr)
+			_G(_unp)._endAdr = _G(_unp)._endAdC;
+
+		_G(_unp)._endAdC = 0;
+		_G(_unp)._fEndAf = 0;
+	}
+
+	if (_G(_unp)._fEndAf && (_G(_unp)._endAdr == 0x10000)) {
+		_G(_unp)._endAdr = r->_mem[_G(_unp)._fEndAf] | r->_mem[_G(_unp)._fEndAf + 1] << 8;
+		if (_G(_unp)._endAdr == 0)
+			_G(_unp)._endAdr = 0x10000;
+		else
+			_G(_unp)._endAdr++;
+		_G(_unp)._fEndAf = 0;
+	}
+
+	if (_G(_unp)._fStrAf /*&&(_G(_unp)._strMem==0x800)*/) {
+		_G(_unp)._strMem = r->_mem[_G(_unp)._fStrAf] | r->_mem[_G(_unp)._fStrAf + 1] << 8;
+		_G(_unp)._strMem++;
+		_G(_unp)._fStrAf = 0;
+	}
+
+	if (_G(_unp)._exoFnd && (_G(_unp)._strMem != 2)) {
+		_G(_unp)._strMem = r->_mem[0xfe] + (r->_mem[0xff] << 8);
+
+		if ((_G(_unp)._exoFnd & 0xff) == 0x30) {
+			_G(_unp)._strMem += r->_y;
+		} else if ((_G(_unp)._exoFnd & 0xff) == 0x32) {
+			_G(_unp)._strMem += r->_y + 1;
+		}
+	}
+
+	if (r->_pc == 0xfce2) {
+		if ((*(unsigned int *)(mem + 0x8004) == 0x38cdc2c3) && (mem[0x8008] == 0x30)) {
+			r->_pc = r->_mem[0x8000] + (r->_mem[0x8001] << 8);
+		}
+	} else if (r->_pc == 0xa7ae) {
+		info->_basicTxtStart = mem[0x2b] | mem[0x2c] << 8;
+		if (info->_basicTxtStart == 0x801) {
+			info->_run = findSys(mem + info->_basicTxtStart, 0x9e);
+			if (info->_run > 0)
+				r->_pc = info->_run;
+		}
+	}
+
+	if (_G(_unp)._wrMemF) {
+		_G(_unp)._wrMemF = 0;
+		for (p = 0x800; p < 0x10000; p += 4) {
+			if (*(unsigned int *)(oldmem + p) == *(unsigned int *)(mem + p)) {
+				*(unsigned int *)(mem + p) = 0;
+				_G(_unp)._wrMemF = 1;
+			}
+		}
+		/* clean also the $fd30 table copy in RAM */
+		if (memcmp(mem + 0xfd30, vector, sizeof(vector)) == 0) {
+			memset(mem + 0xfd30, 0, sizeof(vector));
+		}
+	}
+
+	if (_G(_unp)._lfMemF) {
+		for (p = 0xffff; p > 0x0800; p--) {
+			if (oldmem[--_G(_unp)._lfMemF] == mem[p])
+				mem[p] = 0x0;
+			else {
+				if (p >= 0xffff)
+					_G(_unp)._lfMemF = 0 | _G(_unp)._ecaFlg;
+				break;
+			}
+		}
+	}
+
+	if (*forcedname) {
+		strncpy(name, forcedname, 248);
```

Comment:
g++ also warns here.  Is there a particular reason in limiting to 248 bytes here?  In the `else` below it looks like the 248 limit is to ensure that `".%04x%s"` will have enough space to be appended, but that's not necessary here.  I'd think even `strcpy` would work here, since `name` and `forcedname` are the same size, though to be safer:

```c
snprintf(name, sizeof name, "%s", forcedname);
```
would be better.